odl/solvers/advanced/proximal_operators.py

```diff
@@ -0,0 +1,386 @@
+
+__all__ = ('combine_proximals', 'proximal_zero', 'proximal_nonnegativity',
+           'proximal_convexconjugate_l1', 'proximal_convexconjugate_l2')
+
+# TODO: updated doc
+
+# TODO: remove diagonal op
+def combine_proximals(factory_list):
+
+    def diagonal_operator(operators, dom=None, ran=None):
+        """Broadcast argument to set of operators.
+
+        Parameters
+        ----------
+        operators : array-like
+            An array of `Operator`'s
+        dom : `ProductSpace`, optional
+            Domain of the operator. If not provided, it is tried to be
+            inferred from the operators. This requires each **column**
+            to contain at least one operator.
+        ran : `ProductSpace`, optional
+            Range of the operator. If not provided, it is tried to be
+            inferred from the operators. This requires each **row**
+            to contain at least one operator.
+        """
+
+        indices = [range(len(operators)), range(len(operators))]
+        shape = (len(operators), len(operators))
+        op_matrix = sp.sparse.coo_matrix((operators, indices), shape)
+
+
+        return ProductSpaceOperator(op_matrix, dom=dom, ran=ran)
+
+    def make_diag(step_size):
+
+        return  diagonal_operator(
+            [factory(step_size) for factory in factory_list])
+
+    return make_diag
+
+
+def proximal_zero(space):
+    """Function to create the proximal operator of f(x) = 0.
+
+    Factory function which provides a function to initialize the proximal
+    operator of ``f(x) = 0`` where ``x`` is an element in ``space``. The
+    proximal operator of this functional is the identity operator.
+
+    Parameters
+    ----------
+    space : `DiscreteLp` or `ProductSpace` of `DiscreteLp`
+        Domain of the functional ``f(x)``
+
+    Returns
+    -------
+    make_prox : `function`
+        Function to initialize the proximal operator
+    """
+
+    def make_prox(tau):
+        """Return an instance of the proximal operator.
+
+        Parameters
+        ----------
+        tau : positive `float`
+            Step length parameter. Unused here but introduced to provide a
+            common interface
+
+        Returns
+        -------
+        id : `IdentityOperator`
+            The proximal operator instance of ``f(x) = 0`` which is the
+            identity operator
+        """
+
+        return IdentityOperator(space)
+
+    return make_prox
+
+
+# TODO: update doc
+def proximal_nonnegativity(space):
+    """Function to create the proximal operator of f(x) = 0.
+
+    Factory function which provides a function to initialize the proximal
+    operator of ``f(x)`` where ``x`` is an element in ``space``.
+
+    Parameters
+    ----------
+    space : `DiscreteLp` or `ProductSpace` of `DiscreteLp`
+        Domain of the functional ``f(x)``
+
+    Returns
+    -------
+    make_prox : `function`
+        Function to initialize the proximal operator
+    """
+
+    def make_prox(tau):
```

Comment:
I just came up with a better pattern here:

instead of:

```python
def outer_factor(params):
    def inner_factor(tau):
        class _ProxOp(Operator):
            def __init__(self, params):
                self.params = params
            def _call(x):
                # use self.params and x
        return _ProxOp(tau)
    return inner_factory
```

we should (ofc?) do the following and use `_ProxOp.__init__` as a factory, this avoids one level of indirection.

```python
def outer_factor(params):
    class _ProxOp(Operator):
        def __init__(self, params):
            self.params = params
        def _call(x):
            # use params and x
    return _ProxOp
```